File: leetcode2/0653. Two Sum IV - Input is a BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        self.stack=set()
        
        def subfindtwoSum(root,k):
            if not root:
                return False
            if k-root.val not in self.stack:
                self.stack.add(root.val)
                return subfindtwoSum(root.left,k) or subfindtwoSum(root.right,k)
            else:
                return True
        return subfindtwoSum(root,k)

File: C#/0653. Two Sum IV - Input is a BST.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool FindTarget(TreeNode root, int k) {
        if(root==null){return false;}
        HashSet<int> rootValSet=new HashSet<int>();
        Stack<TreeNode> rootStack=new Stack<TreeNode>();
        rootStack.Push(root);
        while(rootStack.Count()!=0){
            TreeNode curRoot=rootStack.Pop();
            if(rootValSet.Contains(k-curRoot.val)){
                return true;
            }
            rootValSet.Add(curRoot.val);
            if(curRoot.left!=null){
                rootStack.Push(curRoot.left);
            }
            if(curRoot.right!=null){
                rootStack.Push(curRoot.right);
            }
        }
        return false;
    }
}
